From 28079dd6d58046941bd075f5451592fa20c5e0c7 Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Mon, 17 Jun 2024 23:31:32 +0200
Subject: [PATCH 17/23] Add CBaseBlockHeader, separate GetHash and GetPowHash
 (#3)

CBaseBlockHeader is a normal Bitcoin header without auxillary POW data requried for merge-mining. It's also used as a base class for CBlockHeader (hence the name).

This separation is required for when we add auxpow later, to avoid cycles.
---
 src/CMakeLists.txt            |   1 +
 src/primitives/baseheader.cpp |  22 ++++++++
 src/primitives/baseheader.h   |  69 +++++++++++++++++++++++
 src/primitives/block.cpp      |   6 +-
 src/primitives/block.h        |  39 +------------
 src/test/CMakeLists.txt       |   1 +
 src/test/baseheader_tests.cpp | 103 ++++++++++++++++++++++++++++++++++
 7 files changed, 200 insertions(+), 41 deletions(-)
 create mode 100644 src/primitives/baseheader.cpp
 create mode 100644 src/primitives/baseheader.h
 create mode 100644 src/test/baseheader_tests.cpp

diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 5215fe4e6b33c350dfbfe1672dbde81b1af5d3d2..9b17db881d634e5af4fc748cefceabd459b87e3d 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -545,6 +545,7 @@ add_library(common
 	netbase.cpp
 	outputtype.cpp
 	policy/policy.cpp
+	primitives/baseheader.cpp
 	primitives/block.cpp
 	protocol.cpp
 	psbt.cpp
diff --git a/src/primitives/baseheader.cpp b/src/primitives/baseheader.cpp
new file mode 100644
index 0000000000000000000000000000000000000000..e5b00ce0086541f3b80e62f2fb418a89c922d4dd
--- /dev/null
+++ b/src/primitives/baseheader.cpp
@@ -0,0 +1,22 @@
+// Copyright (c) 2024 The Bitcoin developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <primitives/baseheader.h>
+
+#include <chainparams.h>
+#include <crypto/scrypt.h>
+#include <hash.h>
+#include <serialize.h>
+
+BlockHash CBaseBlockHeader::GetHash() const {
+    return BlockHash(SerializeHash(*this));
+}
+
+BlockHash CBaseBlockHeader::GetPowHash() const {
+    CDataStream ss(SER_NETWORK, PROTOCOL_VERSION);
+    ss << *this;
+    uint256 hash;
+    scrypt_1024_1_1_256((const uint8_t *)ss.data(), hash.data());
+    return BlockHash(hash);
+}
diff --git a/src/primitives/baseheader.h b/src/primitives/baseheader.h
new file mode 100644
index 0000000000000000000000000000000000000000..6b62d3d4962024505da01626919507717904a5f2
--- /dev/null
+++ b/src/primitives/baseheader.h
@@ -0,0 +1,69 @@
+// Copyright (c) 2024 The Bitcoin developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_PRIMITIVES_BASE_HEADER_H
+#define BITCOIN_PRIMITIVES_BASE_HEADER_H
+
+#include <primitives/blockhash.h>
+#include <serialize.h>
+#include <uint256.h>
+#include <util/time.h>
+
+/**
+ * Dogecoin specific: A normal Bitcoin header without auxpow information, for
+ * merge-mining.
+ *
+ * This "intermediate step" in constructing the full header is useful, because
+ * it breaks the cyclic dependency between auxpow (referencing a parent block
+ * header) and the block header (referencing an auxpow). The parent block
+ * header does not have auxpow itself, so it is a "base" header.
+ *
+ * In dogecoin, this is called CPureBlockHeader.
+ */
+class CBaseBlockHeader {
+public:
+    // header
+    int32_t nVersion;
+    BlockHash hashPrevBlock;
+    uint256 hashMerkleRoot;
+    uint32_t nTime;
+    uint32_t nBits;
+    uint32_t nNonce;
+
+    CBaseBlockHeader() { SetNull(); }
+
+    SERIALIZE_METHODS(CBaseBlockHeader, obj) {
+        READWRITE(obj.nVersion, obj.hashPrevBlock, obj.hashMerkleRoot,
+                  obj.nTime, obj.nBits, obj.nNonce);
+    }
+
+    void SetNull() {
+        nVersion = 0;
+        hashPrevBlock = BlockHash();
+        hashMerkleRoot.SetNull();
+        nTime = 0;
+        nBits = 0;
+        nNonce = 0;
+    }
+
+    bool IsNull() const { return (nBits == 0); }
+
+    // In Dogecoin, "block hash" and "PoW hash" are two separate hashes:
+
+    /** "Block hash" is using double SHA-256 and used as the unique identifier
+     * of the block, but it doesn't have any PoW done on it. */
+    BlockHash GetHash() const;
+
+    /** "PoW hash" is using Scrypt and miners have to solve this hash to be
+     * below the target. */
+    BlockHash GetPowHash() const;
+
+    NodeSeconds Time() const {
+        return NodeSeconds{std::chrono::seconds{nTime}};
+    }
+
+    int64_t GetBlockTime() const { return (int64_t)nTime; }
+};
+
+#endif // BITCOIN_PRIMITIVES_BASE_HEADER_H
diff --git a/src/primitives/block.cpp b/src/primitives/block.cpp
index 78f35a31326066375071b853663d03aef1c195de..3d1f1ad7763928570633c4fa7f81b0b80d079342 100644
--- a/src/primitives/block.cpp
+++ b/src/primitives/block.cpp
@@ -1,17 +1,13 @@
 // Copyright (c) 2009-2010 Satoshi Nakamoto
 // Copyright (c) 2009-2016 The Bitcoin Core developers
+// Copyright (c) 2024 The Bitcoin developers
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
 #include <primitives/block.h>
 
-#include <hash.h>
 #include <tinyformat.h>
 
-BlockHash CBlockHeader::GetHash() const {
-    return BlockHash{(HashWriter{} << *this).GetHash()};
-}
-
 std::string CBlock::ToString() const {
     std::stringstream s;
     s << strprintf("CBlock(hash=%s, ver=0x%08x, hashPrevBlock=%s, "
diff --git a/src/primitives/block.h b/src/primitives/block.h
index 8fa2d3b1335c73690c7f5f76ae7daa87ce5e1595..c102e15b0abd4a77bd401b729cc6cd598d173a91 100644
--- a/src/primitives/block.h
+++ b/src/primitives/block.h
@@ -1,11 +1,13 @@
 // Copyright (c) 2009-2010 Satoshi Nakamoto
 // Copyright (c) 2009-2016 The Bitcoin Core developers
+// Copyright (c) 2024 The Bitcoin developers
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
 #ifndef BITCOIN_PRIMITIVES_BLOCK_H
 #define BITCOIN_PRIMITIVES_BLOCK_H
 
+#include <primitives/baseheader.h>
 #include <primitives/blockhash.h>
 #include <primitives/transaction.h>
 #include <serialize.h>
@@ -20,42 +22,7 @@
  * the block is a special one that creates a new coin owned by the creator of
  * the block.
  */
-class CBlockHeader {
-public:
-    // header
-    int32_t nVersion;
-    BlockHash hashPrevBlock;
-    uint256 hashMerkleRoot;
-    uint32_t nTime;
-    uint32_t nBits;
-    uint32_t nNonce;
-
-    CBlockHeader() { SetNull(); }
-
-    SERIALIZE_METHODS(CBlockHeader, obj) {
-        READWRITE(obj.nVersion, obj.hashPrevBlock, obj.hashMerkleRoot,
-                  obj.nTime, obj.nBits, obj.nNonce);
-    }
-
-    void SetNull() {
-        nVersion = 0;
-        hashPrevBlock = BlockHash();
-        hashMerkleRoot.SetNull();
-        nTime = 0;
-        nBits = 0;
-        nNonce = 0;
-    }
-
-    bool IsNull() const { return (nBits == 0); }
-
-    BlockHash GetHash() const;
-
-    NodeSeconds Time() const {
-        return NodeSeconds{std::chrono::seconds{nTime}};
-    }
-
-    int64_t GetBlockTime() const { return (int64_t)nTime; }
-};
+class CBlockHeader : public CBaseBlockHeader {};
 
 class CBlock : public CBlockHeader {
 public:
diff --git a/src/test/CMakeLists.txt b/src/test/CMakeLists.txt
index 093e683d9367db7f3f5d4831e9b807eb0b050fa2..b4cbd03f0cde0e54a16fea19f25fc4121728c5d9 100644
--- a/src/test/CMakeLists.txt
+++ b/src/test/CMakeLists.txt
@@ -140,6 +140,7 @@ add_boost_unit_tests_to_suite(bitcoin test_bitcoin
 		base32_tests.cpp
 		base58_tests.cpp
 		base64_tests.cpp
+		baseheader_tests.cpp
 		bip32_tests.cpp
 		bitmanip_tests.cpp
 		blockchain_tests.cpp
diff --git a/src/test/baseheader_tests.cpp b/src/test/baseheader_tests.cpp
new file mode 100644
index 0000000000000000000000000000000000000000..597f792b3a81ced4fef1a05f07db878df25295c2
--- /dev/null
+++ b/src/test/baseheader_tests.cpp
@@ -0,0 +1,103 @@
+// Copyright (c) 2024 The Bitcoin developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <boost/test/unit_test.hpp>
+
+#include <primitives/baseheader.h>
+#include <streams.h>
+#include <util/strencodings.h>
+
+BOOST_AUTO_TEST_SUITE(baseheader_tests)
+
+BOOST_AUTO_TEST_CASE(genesis) {
+    CBaseBlockHeader genesis;
+    genesis.nVersion = 1;
+    genesis.hashPrevBlock = BlockHash();
+    genesis.hashMerkleRoot = uint256S(
+        "5b2a3f53f605d62c53e62932dac6925e3d74afa5a4b459745c36d42d0ed26a69");
+    genesis.nTime = 1386325540;
+    genesis.nBits = 0x1e0ffff0;
+    genesis.nNonce = 99943;
+    BOOST_CHECK_EQUAL(
+        genesis.GetHash().ToString(),
+        "1a91e3dace36e2be3bf030a65679fe821aa1d6ef92e7c9902eb318182c355691");
+    BOOST_CHECK_EQUAL(
+        genesis.GetPowHash().ToString(),
+        "0000026f3f7874ca0c251314eaed2d2fcf83d7da3acfaacf59417d485310b448");
+
+    // Check serialization
+    CDataStream ss(0, 0);
+    ss << genesis;
+    BOOST_CHECK_EQUAL(HexStr(ss),
+                      "01000000000000000000000000000000000000000000000000000000"
+                      "0000000000000000696ad20e2dd4365c7459b4a4a5af743d5e92c6da"
+                      "3229e6532cd605f6533f2a5b24a6a152f0ff0f1e67860100");
+}
+
+BOOST_AUTO_TEST_CASE(block_145000) {
+    CBaseBlockHeader block;
+    block.nVersion = 2;
+    block.hashPrevBlock = BlockHash(uint256S(
+        "919a380db4b45eb97abb131633d87ff690387ebe03ac76690da3f4d681400558"));
+    block.hashMerkleRoot = uint256S(
+        "316614dcd65aa75888cfe1ebb2190740bd8d1fc3e30a0c1952062740b1419c33");
+    block.nTime = 1395094679;
+    block.nBits = 0x1b499dfd;
+    block.nNonce = 1200826624;
+    BOOST_CHECK_EQUAL(
+        block.GetHash().ToString(),
+        "cc47cae70d7c5c92828d3214a266331dde59087d4a39071fa76ddfff9b7bde72");
+    BOOST_CHECK_EQUAL(
+        block.GetPowHash().ToString(),
+        "00000000002dfb87dd0d1b359eac948f33e91f87d586d36c497df2b08db7eb8a");
+
+    // Check serialization
+    CDataStream ss(0, 0);
+    ss << block;
+    BOOST_CHECK_EQUAL(HexStr(ss),
+                      "0200000058054081d6f4a30d6976ac03be7e3890f67fd8331613bb7a"
+                      "b95eb4b40d389a91339c41b140270652190c0ae3c31f8dbd400719b2"
+                      "ebe1cf8858a75ad6dc14663197742753fd9d491b00299347");
+}
+
+BOOST_AUTO_TEST_CASE(block_371336) {
+    // Block before the first merge-mined block
+    CBaseBlockHeader block;
+    block.nVersion = 0x00620002;
+    block.hashPrevBlock = BlockHash(uint256S(
+        "8ad58fc406423207bdd82bed27c0c9a22f8241e3d3e8595191decb55a50b20c7"));
+    block.hashMerkleRoot = uint256S(
+        "a0503bb44fd98e79239cc8f7b896a81b9a5fb1deb74e165173d7a530db34d877");
+    block.nTime = 1410464569;
+    block.nBits = 0x1b2fdf75;
+    block.nNonce = 3401887720;
+    BOOST_CHECK_EQUAL(
+        block.GetHash().ToString(),
+        "46a8b109fb016fa41abd17a19186ca78d39c60c020c71fcd2690320d47036f0d");
+    BOOST_CHECK_EQUAL(
+        block.GetPowHash().ToString(),
+        "00000000000d07e96b781d6336d1badd7fc64ade5fd5957c596475a893d9b763");
+}
+
+BOOST_AUTO_TEST_CASE(block_371337) {
+    // First merge-mined block
+    CBaseBlockHeader block;
+    block.nVersion = 0x00620102;
+    block.hashPrevBlock = BlockHash(uint256S(
+        "46a8b109fb016fa41abd17a19186ca78d39c60c020c71fcd2690320d47036f0d"));
+    block.hashMerkleRoot = uint256S(
+        "ee27b8fb782a5bfb99c975f0d4686440b9af9e16846603e5f2830e0b6fbf158a");
+    block.nTime = 1410464577;
+    block.nBits = 0x1b364184;
+    block.nNonce = 0;
+    BOOST_CHECK_EQUAL(
+        block.GetHash().ToString(),
+        "60323982f9c5ff1b5a954eac9dc1269352835f47c2c5222691d80f0d50dcf053");
+    // Merge-mined block has no PoW done on the header itself
+    BOOST_CHECK_EQUAL(
+        block.GetPowHash().ToString(),
+        "2486dafe34a0258425fbf7dd0c63b70f10c5803db63e9a61a1af5d2a2fc39146");
+}
+
+BOOST_AUTO_TEST_SUITE_END()
